src/notebooks/phdi_analytics.py

```diff
@@ -0,0 +1,297 @@
+        addresses.append(street)
+        zip_codes.append(random.sample(FAIRFAX_COUNTY_ZIPS, 1)[0])
+        event_dates.append(generate_random_date(start_date, end_date))
+
+    df = pd.DataFrame({
+        'pprl_generated_id': personids,
+        'first_name': first_names,
+        'last_name': last_names,
+        'street_address': addresses,
+        'zip_code': zip_codes,
+        'latitude': latitudes,
+        'longitude': longitudes,
+        'event_dt': event_dates,
+    })
+
+    with ThreadPoolExecutor() as tpe:
+        data = list(
+            tqdm(
+                tpe.map(geocode, df[['latitude', 'longitude']].itertuples()),
+                total=len(df)
+            )
+        )
+    data_df = pd.DataFrame.from_records(data)
+
+    df['census_tract'] = data_df['state'] + data_df['county'] + data_df['tract']
+    df['census_block_group'] = data_df['state'] +\
+        data_df['county'] + data_df['tract'] + data_df['block'].str[0]
+    df['census_block'] = data_df['geoid']
+
+    is_fully_vax = random.choices([0, 1], weights=(0.4, 0.6), k=len(points))
+    df['is_fully_vax'] = is_fully_vax
+    df['tested_covid_positive'] = df.apply(
+        lambda row: generate_covid_positive(row), 1)
+    df['breakthrough_infection'] = df['is_fully_vax'] *\
+        df['tested_covid_positive']
+
+    df['event_dt'] = pd.to_datetime(df['event_dt'])
+
+    # Write as Pickle File
+    df.to_pickle(os.path.join(OUTPUTS_DIR, 'synthetic_df.pkl'))
+
+    return df
+
+
+def aggregate_df(df, window='30D', metric_label='Breakthrough Infection'):
+    df.sort_values(by='event_dt', ascending=True, inplace=True)
+
+    if metric_label.lower() == 'breakthrough infection':
+        metric = 'breakthrough_infection'
+
+    if metric_label.lower() == 'covid infection':
+        metric = 'tested_covid_positive'
+
+    df['rolling_sum'] = df.rolling(window, on='event_dt')[metric].sum()
+    df['rolling_count'] = df.rolling(window, on='event_dt')[metric].count()
+    df['rolling_rate'] = df['rolling_sum'] / df['rolling_count']
+
+    return df
+
+
+def plot_time_series_metric(df, window, metric_label):
+    df_agg = aggregate_df(df, window=window, metric_label=metric_label)
+
+    window_length_label = re.findall(r'\d+', window)[0]
+
+    fig, ax_lst = plt.subplots(nrows=1, ncols=1, figsize=(15, 8))
+
+    sns.lineplot(
+        data=df_agg,
+        x='event_dt',
+        y='rolling_rate',
+        ax=ax_lst,
+        ci=None
+    )
+
+    ax_lst.set_ylabel(metric_label, fontsize=14)
+    ax_lst.set_xlabel('Date', fontsize=14)
+
+    ax_lst.set_title('{} Rate ({} Day Rolling)'.format(
+        metric_label,
+        window_length_label
+    ), fontsize=22)
+    ax_lst.yaxis.set_major_formatter(mtick.PercentFormatter(1.0))
+
+    plt.tight_layout(h_pad=6)
+
+    return fig, ax_lst
+
+
+def create_map(geography, shapefile_path, metric, df):
+    """
+    Create spatial analytics map
+    """
+    shapefile_map = gpd.read_file(shapefile_path)
```

Comment:
The dependency on `geopandas` is a nightmare on Windows. In order to use `geopandas`, you have to have the `GDAL` API installed, and in order to install that, you have to have the Microsoft Visual C++ 14.0 or later build tools installed. Is there another way to read shapefiles? I remember dealing with these issues in R back when I was doing spatial analytics and it was a nightmare. I'm hoping an alternative has come out that helps alleviate this. If not, no worries.

Reply:
Yeah agreed. I am on Windows, but I'm actually using Anaconda to manage my environment (which I know is inconsistent with the rest of the engineering team and prototype). For this spatial analytics notebook, I think using `geopandas` is the way to go because it really is the simplest way IMO to work with shapefiles in Python / Pandas

Reply:
Agreed. You've got it up and running and it produces the results we need. Plus GDAL seems to be the standard for spatial analytics. But wow, what a nightmare on Windows.